### PR TITLE
Eventbrite Block: Update Button

### LINKED ISSUE #15710: Eventbrite: Replace with SubmitButton new Button block
#### Is your feature request related to a problem? Please describe.

Now we have a [shared button block](https://github.com/Automattic/jetpack/pull/15370), we should update the Eventbrite block to use this instead of the SubmitButton component.

From https://github.com/Automattic/jetpack/issues/14761

### PR DESCRIPTION
**Pull request by Copons**
Fixes #15710

#### Changes proposed in this Pull Request:

* Replaces the Eventbrite block's `ModalButtonPreview` component with the `Button` inner block.
* Update the Button block to support fill and outline styles, and inline formats (required by Eventbrite and others).

| Before | After |
| - | - |
| Screenshot 2020-05-25 at 15 42 54 | Screenshot 2020-05-25 at 15 42 25 Screenshot 2020-05-25 at 15 54 37 |

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

* Before checking out this PR:
  * Insert two Eventbrite block, and embed an event for both of them.
  * Change the style of one of the block to "Button & Modal", while leaving the other as "In-page Embed".
  * Change the "Button & Modal" block's settings as much as possible: colors, border radius, text, style.
  * Save.
* Checkout this PR, and reload the post.
* Look at the browser console. Make sure that only one of the two blocks (the "Button & Modal" one) has migrated.
* Make sure the block and all the customizations have migrated correctly.
* Save and preview.
* Make sure the button works as expected: clicking it will open an Eventbrite modal (over HTTPS only, otherwise it will simply opens an Eventbrite page in a new tab).

#### Proposed changelog entry for your changes:

* Eventbrite block: Use new Button inner block.
* Button block: Add support for button styles and inline formats.

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
This all tests well for me. 👍

**Comment by leogermani** [comment]
Hi,

I've tested this PR and it worked fine in my local environment. Only two small observations:

* I couldn't confirm that "that only one of the two blocks (the "Button & Modal" one) has migrated." I don't see anything related to that in the console.
* When you add a new block and embed the event, the "Embed Type" option doesn't show neither of the options selected, which is a bit confusing (this might be a problem that was there before and can be addressed in a separate PR, but I thought it was worth reporting)

But testing with the Gutenberg plugin I saw some strange behaviors

When I add the block, and then change it to "Button & Modal", the "Embed Type" section disappears, and the title of the sidebar changes to "Button".

The button controls work fine, but if I click to edit another block and then go back to this block, I only see the "Embed Type" section and I can't get to the button options again.

Can you reproduce that?

**Comment by Copons** [comment]
> Hi,
> 
> I've tested this PR and it worked fine in my local environment. Only two small observations:
> 
> * I couldn't confirm that "that only one of the two blocks (the "Button & Modal" one) has migrated." I don't see anything related to that in the console.

You should see something like this on block migration:

Screenshot 2020-06-11 at 15 36 23

This should only happen for Eventbrite blocks in "Button & Modal" style, since otherwise there wouldn't be anything to migrate.
Once migrated, if you save (or autosave) the post, the updated block will "persist", so reloading again won't trigger the migration anymore.

> * When you add a new block and embed the event, the "Embed Type" option doesn't show neither of the options selected, which is a bit confusing (this might be a problem that was there before and can be addressed in a separate PR, but I thought it was worth reporting)
> 
> But testing with the Gutenberg plugin I saw some strange behaviors
> 
> When I add the block, and then change it to "Button & Modal", the "Embed Type" section disappears, and the title of the sidebar changes to "Button".
> 
> The button controls work fine, but if I click to edit another block and then go back to this block, I only see the "Embed Type" section and I can't get to the button options again.
> 
> Can you reproduce that?

This sounds like the correct behaviour, which is a little awkward for blocks containing only the Button inner block.

Basically, if you click on the button, you enter in the Button block context.
If you click outside the button, but still inside the parent block, you enter in the Eventbrite block context.

Jun-11-2020 15-42-08

**Comment by leogermani** [comment]
@Copons thanks for the explanation,

Now that I understand what was going on on Gutenberg (which is bit confusing) I know why I don't see the upgrade message on my console. That's because the block is not being upgraded...

Here's a detailed step by step of what I'm doing

* Running Jetpack locally using the docker environment and ngrok
* checkout master branch
* yarn build-client
* Create a post with 2 blocks as described
* Click "save draft"
* Checkout this branch
* yarn build-client
* refresh the editor
* Nothing on the console and the block is still the same (no Button block inside the eventbritte block)

I actually tried both running `build-client` and not running it. So I'm not sure what I'm missing.

**Comment by Copons** [comment]
> * yarn build-client

Ha! Found the culprit, @leogermani!
Blocks and other block editor extensions are built in a different bundle than `client`, which contains the JS for the Jetpack dashboard.
You need to run `yarn build-extensions` (add a `--watch` if you want to keep it rebuilding on changes). 🙂

File: extensions/blocks/eventbrite/eventbrite.php [hunk 1]
```diff
@@ -117,6 +117,9 @@ function render_block( $attr, $content ) {
 	);
 
 	// Replace the placeholder id saved in the post_content with a unique id used by widget.js.
+	$content = str_replace( 'eventbrite-widget-id', $widget_id, $content );
```
Comment (fgiannar) [inline extensions/blocks/eventbrite/eventbrite.php:120]:
```suggestion
	$content = str_replace( 'eventbrite-widget-id', esc_attr( $widget_id ), $content );
```

File: extensions/blocks/eventbrite/eventbrite.php [hunk 2]
```diff
@@ -117,6 +117,9 @@ function render_block( $attr, $content ) {
...
+
+	// Fallback for block version deprecated/v2.
 	$content = preg_replace( '/eventbrite-widget-\d+/', $widget_id, $content );
```
Comment (fgiannar) [inline extensions/blocks/eventbrite/eventbrite.php:123]:
```suggestion
	$content = preg_replace( '/eventbrite-widget-\d+/', esc_attr( $widget_id ), $content );
```

**Comment by Copons** [comment]
> Hi there,
> 
> While testing different styles/settings I had the following issue:
> 
> * Added a "Button & Modal" style Eventbrite block
> * Aligned it right
> * The block element seems to be floating right, therefore couldn't display the Eventbrite block context anymore by clicking anywhere around it.

This is a known issue: https://github.com/Automattic/jetpack/issues/15580
I think it should be solved elsewhere though! 🙂

**Comment by Copons** [comment]
In 9296d7d I've set the `inline` embed type as default, so that it will be selected upon first embed.

I've also updated the edit render function to only display the embed type controls when the embed URL resolved correctly (aka: when the block shows the actual preview).
Previously, it was displayed always unless the URL couldn't be embedded. Meaning, it was visible when the block was in placeholder mode and even in "embedding". Though, changing the setting while in those modes isn't possible, so there's no reason to show those controls at that point.

Thanks for pointing this out @leogermani and @fgiannar, I honestly never noticed it!

**Comment by leogermani** [comment]
@Copons thanks for all the explanations and sorry not building this properly since day one!

It looks good for me now!

(tip: when you answer anything, feel free to remove the "needs author reply" label and add "needs review" again. This will make the PR show up in our queue faster again)

**Comment by Copons** [comment]
Waiting until release day (July 7) to merge D45128-code.

**Comment by Copons** [comment]
r210104-wpcom
